File: ds/linked_list/2_146.lru-cache/146.lru-cache.py
#
# @lc app=leetcode id=146 lang=python3
#
# [146] LRU Cache
#

# @lc code=start
from collections import OrderedDict

class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity             # Maximum number of items the cache can hold
        self.cache = OrderedDict()          # Stores the cache items in access order

    def get(self, key: int) -> int:
        if not key in self.cache:           # If key not present, return -1
            return -1 
        self.cache.move_to_end(key)           # Mark key as recently used (move to end)
        return self.cache[key]               # Return the value for the key

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.cache.move_to_end(key)         # If key exists, mark as recently used
        self.cache[key] = value                 # Insert or update the key-value pair
        if len(self.cache) > self.capacity:    # If over capacity,
            self.cache.popitem(False)           # Remove least recently used item (from front)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
# @lc code=end

File: ds/linked_list/2_146.lru-cache/146.lru-cache.cpp
/*
 * @lc app=leetcode id=146 lang=cpp
 *
 * [146] LRU Cache
 */

// @lc code=start
#include <unordered_map>

class LRUCache {
    struct Node{
        int key; // mapping to remove node
        int value;
        Node* left;
        Node* right;
        Node(int key, int value) : key(key), value(value){}
    };

    Node* head;
    Node* tail;
    std::unordered_map<int, Node*> cache;
    int capacity;

    void move_to_end(Node* curr){
        auto before = tail->left;
        before->right = curr;
        curr->left = before;
        curr->right = tail; 
        tail->left = curr; 
    }

    void rewire_pointers(Node* curr){
        auto before = curr->left;
        auto after = curr->right;
        before->right=after;
        after->left=before;
    }


    public:
    LRUCache(int capacity) : capacity(capacity){
       head = new Node(-1,-1);
       tail = new Node(-1,-1);
        head->right = tail;
        tail->left = head;
    }

    int get(int key) {
        if (!cache.contains(key)){
            return -1;
        }
        Node* curr = cache[key];
        int val = curr->value;
        rewire_pointers(curr);
        move_to_end(curr);
        return val;

    }
    
    void put(int key, int value) {


        if (cache.contains(key)){
            Node* curr = cache[key];
            rewire_pointers(curr);
            //curr->value = value;
            cache.erase(key);
        }
            Node* inserted = new Node(key, value);
            cache[key]=inserted;
            move_to_end(inserted);

            if (cache.size()> capacity){
                Node* curr = head->right;
                rewire_pointers(curr);
                //move_to_end(curr);
                cache.erase(curr->key);
                //delete curr;
        }
    }
};

// single linked list doesnt have ref to prev node
// list is expensive
// dll is a possible solution 

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
// @lc code=end